src/query/server/query.go

```diff
@@ -643,18 +644,41 @@ func Run(runOpts RunOptions) RunResult {
 		}
 	}
 
-	prometheusEngine, err := newPromQLEngine(cfg, prometheusEngineRegistry,
-		instrumentOptions)
+	defaultPrometheusEngine, err := newPromQLEngine(lookbackDuration, cfg, prometheusEngineRegistry, instrumentOptions)
 	if err != nil {
 		logger.Fatal("unable to create PromQL engine", zap.Error(err))
 	}
+	prometheusEngineFn := func(lookbackDuration time.Duration) (*prometheuspromql.Engine, error) {
+		// NB: use nil metric registry to avoid duplicate metric registration when creating multiple engines
+		return newPromQLEngine(lookbackDuration, cfg, nil, instrumentOptions)
+	}
+
+	enginesByLookback := make(map[time.Duration]*prometheuspromql.Engine)
+	enginesByLookback[lookbackDuration] = defaultPrometheusEngine
+	if mult := cfg.Middleware.Prometheus.ResolutionMultiplier; mult > 0 {
```

Comment:
nit: could we extract this to a function, maybe next to query.go and unit test this properly? query.go is a huge file already and is hard to navigate sometimes.

Reply:
https://github.com/m3db/m3/pull/3882/commits/eadb0bd569b2c903058feb6f49cbc96bf9ba0e99